src/transaction/transaction.rs

```diff
@@ -497,7 +499,10 @@ impl Transaction {
             .map(|it: Key| it.into())
             .collect();
         keys.sort();
-        let primary_lock = keys[0].clone();
+        let primary_lock = self
+            .primary_key
+            .get_or_insert_with(|| keys[0].clone())
+            .clone();
```

Comment:
This is a bit odd because it is the only place we set the primary key, so optimistic transactions will not get a primary key, which they should. As I understand it, we already deal with primary keys in the buffer, so I think here we should use that primary key (or keys[0] if it is None).

Reply:
> This is a bit odd because it is the only place we set the primary key, so optimistic transactions will not get a primary key, which they should. As I understand it, we already deal with primary keys in the buffer, so I think here we should use that primary key (or keys[0] if it is None).

After some thought, I'm afraid currently `Buffer` can't handle this right. 
`Buffer` use a `BTreeMap` to store mutations, and regard the minium key in it as the primary key (see `to_proto_mutations` and `commit_primary`), but for example:
- We lock ["b", "c"] first,  since buffer is empty now, we'll use "b" as primary key and lock them.
- Then we want to lock ["a"], this seems work well, but after lock and "a" put into the buffer, "a" become the primary key.
- Then other lock requests will regard "a" as primary key, which is different with the first lock request.
So it seems we'll still need a `primary_key` field, maybe I'll put it in the `Buffer` struct.